pages/cookbook/primitive-types.mdx

```diff
@@ -0,0 +1,243 @@
+```
+
+<Callout>
+
+  **Useful links:**\
+  [`Cell.hash(){:tact}` in Language→Reference](/language/ref/cells#cellhash)
+
+</Callout>
+
+## `Slice`
+
+### How to determine if a `Slice` is empty
+
+A [`Slice{:tact}`](/book/types#primitive-types) is considered _empty_ if it has no stored `data` **and** no stored `references`.
+
+Use `empty(){:tact}` method to check if a [`Slice{:tact}`](/book/types#primitive-types) is empty.
+
+```tact
+// Create an empty Slice with no data and no refs
+let empty_slice: Slice = emptyCell().asSlice();
+// Returns `true`, because `empty_slice` doesn't have any data or refs
+empty_slice.empty();
+
+// Create a slice with some data
+let slice_with_data: Slice = beginCell().
+    storeUint(42, 8).
+    asSlice();
+// Returns `false`, because the slice has some data
+slice_with_data.empty();
+
+// Create a slice with a reference to an empty cell
+let slice_with_refs: Slice  = beginCell().
+    storeRef(emptyCell()).
+    asSlice();
+// Returns `false`, because the slice has a reference
+slice_with_refs.empty();
+
+// Create a slice with data and a reference
+let slice_with_data_and_refs: Slice  = beginCell().
+    storeUint(42, 8).
+    storeRef(emptyCell()).
+    asSlice();
+// Returns `false`, because the slice has both data and reference
+slice_with_data_and_refs.empty();
+```
+
+### How to determine if a `Slice` has no refs (but may have bits)
+
+```tact
+let slice_with_data: Slice = beginCell().
+    storeUint(0, 1).
+    asSlice(); // create a slice with data but without refs
+let refsCount: Int = slice_with_data.refs(); // 0
+let hasNoRefs: Bool = slice_with_data.refsEmpty(); // true
+```
+
+### How to determine if a `Slice` has no data (no bits, but may have refs)
+
+```tact
+let slice_with_data: Slice = beginCell().
+    storeRef(emptyCell()).
+    asSlice(); // create a slice with ref but without data
+let bitsCount: Int = slice_with_data.bits(); // 0
+let hasNoData: Bool = slice_with_data.dataEmpty(); // true
+```
+
+### How to determine if two Slices are equal
+
+Direct comparison:
+
+```tact
+let firstSlice: Slice = "A".asSlice();
+let secondSlice: Slice = "A".asSlice();
+
+let areEqual: Bool = firstSlice == secondSlice;
+let areNotEqual: Bool = firstSlice != secondSlice;
+
+dump(areEqual) // true;
+dump(areNotEqual) // false;
+```
+
+Note, that direct comparison via `=={:tact}` or `!={:tact}` operators implicitly uses hashes under the hood.
+
+Explicit comparisons using `.hash(){:tact}` are also available:
+
+```tact
+fun areSlicesEqual(a: Slice, b: Slice): Bool {
+    return a.hash() == b.hash();
+}
+
+let firstSlice: Slice = "A".asSlice();
+let secondSlice: Slice = "A".asSlice();
+
+let result: Bool = areSlicesEqual(firstSlice, secondSlice);
+dump(result) // true;
+```
+
+<Callout>
+
+  **Useful links:**\
+  [`String.asSlice(){:tact}` in Language→Reference](/language/ref/strings#stringasslice)
+
+</Callout>
+
+## `Int` ↔ `String` [#int-and-string]
+
+### How to convert a `String` to an `Int`
```

Comment:
this one we don't need to move to Book, because it looks kind of like a recipe (although to make it self-contained, we need to create a function of type `String -> Int?`  with this functionality)

Reply:
Will do. Also: https://github.com/tact-lang/tact-docs/issues/168#issuecomment-2037001102